# Opdracht "Solar": Tips & tricks

## Doelstelling

- Je ontwikkelt een machine learning model dat op basis van opbrengsten uit het verleden, waarnemingen van het weer en de tijdstippen van zonsopgang en -ondergang de opbrengst (in kWh) van een individuele zonnepaneelinstallatie kan voorspellen voor de komende uren (opbrengst per uur in kWh).  
- Het model wordt opgesteld aan de hand van waarnemingen uit het verleden, maar voor de komende uren hebben we die natuurlijk niet. We werken daarom met weersvoorspellingen wanneer het model in productie gebruikt wordt. 
- Je ontwikkelt naast het model ook een Python-toepassing waarin je het model gebruikt. Deze toepassing krijgt als input een nieuwe forecast-file voor de komende 48 u. Deze file ziet er als volgt uit: 


In [2]:
import pandas as pd

forecast = pd.read_csv('datasets/forecast.csv')
forecast

timestamp   temp  pressure  cloudiness  humidity_relative
0   2024-05-22T10:00:00  14.34       988        4.72                 85
1   2024-05-22T11:00:00  14.57       992        4.88                 84
2   2024-05-22T12:00:00  14.99       996        5.36                 82
3   2024-05-22T13:00:00  16.24       999        6.00                 76
4   2024-05-22T14:00:00  18.02      1003        6.00                 64
5   2024-05-22T15:00:00  18.60      1008        2.96                 62
6   2024-05-22T16:00:00  17.81      1008        4.88                 68
7   2024-05-22T17:00:00  17.38      1009        4.72                 72
8   2024-05-22T18:00:00  16.92      1009        4.08                 70
9   2024-05-22T19:00:00  16.55      1010        3.36                 72
10  2024-05-22T20:00:00  15.59      1010        2.88                 76
11  2024-05-22T21:00:00  14.03      1010        0.32                 81
12  2024-05-22T22:00:00  12.73      1011        0.32                 86
13  2024-05-22T23:00:00  11.64      1012        0.24                 92
14  2024-05-23T00:00:00  11.16      1012        0.32                 93
15  2024-05-23T01:00:00  10.61      1012        0.32                 96
16  2024-05-23T02:00:00  10.36      1012        0.24                 97
17  2024-05-23T03:00:00  10.28      1012        0.80                 98
18  2024-05-23T04:00:00  10.32      1012        2.32                 98
19  2024-05-23T05:00:00  11.38      1013        3.44                 92
20  2024-05-23T06:00:00  11.67      1013        3.36                 91
21  2024-05-23T07:00:00  11.93      1013        3.04                 93
22  2024-05-23T08:00:00  13.06      1014        2.64                 89
23  2024-05-23T09:00:00  14.48      1014        4.08                 84
24  2024-05-23T10:00:00  15.43      1015        5.04                 76
25  2024-05-23T11:00:00  16.42      1015        4.80                 72
26  2024-05-23T12:00:00  16.54      1015        5.36                 72
27  2024-05-23T13:00:00  16.57      1015        5.60                 72
28  2024-05-23T14:00:00  17.10      1015        5.60                 67
29  2024-05-23T15:00:00  17.17      1015        6.40                 67
30  2024-05-23T16:00:00  17.27      1016        6.48                 67
31  2024-05-23T17:00:00  17.20      1016        5.76                 66
32  2024-05-23T18:00:00  16.70      1016        5.36                 69
33  2024-05-23T19:00:00  16.59      1016        5.44                 67
34  2024-05-23T20:00:00  16.23      1016        5.04                 72
35  2024-05-23T21:00:00  14.24      1017        1.28                 81
36  2024-05-23T22:00:00  12.42      1017        0.80                 86
37  2024-05-23T23:00:00  11.68      1018        0.72                 89
38  2024-05-24T00:00:00  11.13      1018        0.56                 91
39  2024-05-24T01:00:00  10.61      1018        0.56                 92
40  2024-05-24T02:00:00  10.15      1018        0.40                 93
41  2024-05-24T03:00:00   9.60      1018        0.24                 94
42  2024-05-24T04:00:00   9.25      1018        0.56                 94
43  2024-05-24T05:00:00   8.91      1018        1.04                 94
44  2024-05-24T06:00:00   8.85      1018        1.68                 94
45  2024-05-24T07:00:00   9.99      1019        1.76                 91
46  2024-05-24T08:00:00  12.00      1019        1.44                 84
47  2024-05-24T09:00:00  13.67      1020        2.56                 79

Tijdens de eindpresentatie zal je een update van deze file ontvangen, die live zal gegenereerd worden voor de komende 48u.  

Je applicatie zal vervolgens, gebruik makend van je model, een voorspelling genereren voor de komende 48u voor de opbrengst per uur. Hieronder zie je een voorbeeld van de output (als de voorspelling opgevraagd wordt tussen 11 en 12u). 

Voorspelling voor 11u:  0.27 kWh
Voorspelling voor 12u:  0.27 kWh
Voorspelling voor 13u:  0.28 kWh
Voorspelling voor 14u:  0.3 kWh
Voorspelling voor 15u:  0.34 kWh
Voorspelling voor 16u:  0.29 kWh
Voorspelling voor 17u:  0.22 kWh
Voorspelling voor 18u:  0.19 kWh
Voorspelling voor 19u:  0.02 kWh
Voorspelling voor 20u:  0.0 kWh
Voorspelling voor 21u:  0.0 kWh
Voorspelling voor 22u:  0.0 kWh
Voorspelling voor 23u:  0.0 kWh
Voorspelling voor 0u:   0.0 kWh
Voorspelling voor 1u:   0.0 kWh
Voorspelling voor 2u:   0.0 kWh
Voorspelling voor 3u:   0.0 kWh
Voorspelling voor 4u:   0.0 kWh
Voorspelling voor 5u:   0.0 kWh
Voorspelling voor 6u:   0.02 kWh
Voorspelling voor 7u:   0.16 kWh
Voorspelling voor 8u:   0.26 kWh
Voorspelling voor 9u:   0.6 kWh
Voorspelling voor 10u:  0.76 kWh
Voorspelling voor 11u:  0.94 kWh
Voorspelling voor 12u:  0.85 kWh
Voorspelling voor 13u:  1.0 kWh
Voorspelling voor 14u:  0.86 kWh
Voorspelling voor 15u:  0.7 kWh
Voorspelling voor 16u:  0.72 kWh
Voorspelling voor 17u:  0.43 kWh
Voorspelling voor 18u:  0.21 kWh
Voorspelling voor 19u:  0.02 kWh
Voorspelling voor 20u:  0.0 kWh
Voorspelling voor 21u:  0.0 kWh
Voorspelling voor 22u:  0.0 kWh
Voorspelling voor 23u:  0.0 kWh
Voorspelling voor 0u:   0.0 kWh
Voorspelling voor 1u:   0.0 kWh
Voorspelling voor 2u:   0.0 kWh
Voorspelling voor 3u:   0.0 kWh
Voorspelling voor 4u:   0.0 kWh
Voorspelling voor 5u:   0.0 kWh
Voorspelling voor 6u:   0.02 kWh
Voorspelling voor 7u:   0.16 kWh
Voorspelling voor 8u:   0.26 kWh
Voorspelling voor 9u:   0.53 kWh
Voorspelling voor 10u:  0.63 kWh

## Enkele tips voor de data-preparatie

- Je moet drie datasets aan elkaar koppelen tot één geheel. Let er op dat je dat op de juiste manier doet: 
  - Koppel op de juiste velden.
  - Kijk na hoeveel lijnen je overhoudt na koppeling en interpreteer dit aantal. Kan het kloppen? 
  - Lees goed de beschikbare documentatie i.v.m. de weergegevens (weather.csv) en de zonnepanelen (solar.csv).
- Gebruik de laatste versies van solar.csv en weather.csv die je vindt op Chamilo. Hoe meer data, hoe beter voor je model. 
- Gebruik je gezond verstand om te bepalen welke attributen belangrijk kunnen zijn voor de opbrengst van zonnepanelen. Aarzel niet om zelf berekende attributen toe te voegen. 
- Maak enkel gebruik van de technieken voor data-preparatie en modellen die gezien zijn in de cursus. Ga dus niet op zoek naar geavanceerde technieken voor tijdsreeksen. Die heb je niet nodig om een goed voorspellingsmodel te kunnen opstellen. 
- Data voorbereiden en exploreren, modellen fine-tunen en interpreteren doe je in een Python-notebook, zodat je tijdens de presentatie gemakkelijk kan tonen welke stappen je doorlopen hebt en welke outputs je kreeg (er zal geen tijd zijn om te hertrainen). 
- Zoals elke IOT-toepassing kan de input data fouten bevatten, zoals ontbrekende gegevens en outliers. Spoor die op en zoek een verantwoorde manier om ermee om te gaan. 